runtime/query.go

```diff
@@ -45,362 +46,294 @@ func (*defaultQueryParser) Parse(msg proto.Message, values url.Values, filter *u
-		isLast := i == len(fieldPath)-1
-		if !isLast && m.Kind() != reflect.Struct {
-			return fmt.Errorf("non-aggregate type in the mid of path: %s", strings.Join(fieldPath, "."))
+		fields := msgValue.Descriptor().Fields()
+
+		// Get field by name
+		fieldDesc = fields.ByName(protoreflect.Name(fieldName))
+		if fieldDesc == nil {
+			fieldDesc = fields.ByJSONName(fieldName)
+			if fieldDesc == nil {
+				grpclog.Infof("field not found in %q: %q", msgValue.Descriptor().FullName(), strings.Join(fieldPath, "."))
+				return nil
+			}
 		}
-		var f reflect.Value
-		var err error
-		f, props, err = fieldByProtoName(m, fieldName)
-		if err != nil {
-			return err
-		} else if !f.IsValid() {
-			grpclog.Infof("field not found in %T: %s", msg, strings.Join(fieldPath, "."))
-			return nil
+
+		// If this is the last element, we're done
+		if i == len(fieldPath)-1 {
+			break
 		}
 
-		switch f.Kind() {
-		case reflect.Bool, reflect.Float32, reflect.Float64, reflect.Int32, reflect.Int64, reflect.String, reflect.Uint32, reflect.Uint64:
-			if !isLast {
-				return fmt.Errorf("unexpected nested field %s in %s", fieldPath[i+1], strings.Join(fieldPath[:i+1], "."))
-			}
-			m = f
-		case reflect.Slice:
-			if !isLast {
-				return fmt.Errorf("unexpected repeated field in %s", strings.Join(fieldPath, "."))
-			}
-			// Handle []byte
-			if f.Type().Elem().Kind() == reflect.Uint8 {
-				m = f
-				break
-			}
-			return populateRepeatedField(f, values, props)
-		case reflect.Ptr:
-			if f.IsNil() {
-				m = reflect.New(f.Type().Elem())
-				f.Set(m.Convert(f.Type()))
-			}
-			m = f.Elem()
-			continue
-		case reflect.Struct:
-			m = f
-			continue
-		case reflect.Map:
-			if !isLast {
-				return fmt.Errorf("unexpected nested field %s in %s", fieldPath[i+1], strings.Join(fieldPath[:i+1], "."))
-			}
-			return populateMapField(f, values, props)
+		switch fieldDesc.Kind() {
+		case protoreflect.MessageKind, protoreflect.GroupKind:
 		default:
-			return fmt.Errorf("unexpected type %s in %T", f.Type(), msg)
+			return fmt.Errorf("invalid path: %q is not a message", fieldName)
+		}
+
+		// Get the nested message
+		msgValue = msgValue.Mutable(fieldDesc).Message()
+	}
+
+	// Check if oneof already set
+	if of := fieldDesc.ContainingOneof(); of != nil {
+		if f := msgValue.WhichOneof(of); f != nil {
+			return fmt.Errorf("field %q already set for oneof %q", f.FullName().Name(), of.FullName().Name())
 		}
 	}
-	switch len(values) {
-	case 0:
-		return fmt.Errorf("no value of field: %s", strings.Join(fieldPath, "."))
-	case 1:
-	default:
-		grpclog.Infof("too many field values: %s", strings.Join(fieldPath, "."))
+
+	switch {
+	case fieldDesc.IsList():
+		return populateRepeatedField(fieldDesc, msgValue, values)
+	case fieldDesc.IsMap():
+		return populateMapField(fieldDesc, msgValue, values)
+	}
+
+	if len(values) > 1 {
+		grpclog.Infof("too many field values for %q: %q", fieldDesc.FullName().Name(), strings.Join(fieldPath, "."))
```

Comment:
This error seems a bit redundant. It gives the short name of the current field and also the full fieldPath, where the last segment of the fieldPath should be more or less identical to the name of the current field.

Reply:
Yeah, this was wrong, I've changed the wording and this and made it an error. Thanks!